Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+structure IsCharTwoJZeroNF : Prop where
+  a₁ : W.a₁ = 0
+  a₂ : W.a₂ = 0
+
+namespace IsCharTwoJZeroNF
+
+variable {W} (self : W.IsCharTwoJZeroNF)
+include self
+
+theorem b₂ : W.b₂ = 0 := by
+  rw [WeierstrassCurve.b₂, self.a₁, self.a₂]
+  ring1
+
+theorem b₄ : W.b₄ = 2 * W.a₄ := by
+  rw [WeierstrassCurve.b₄, self.a₁]
+  ring1
+
+theorem b₈ : W.b₈ = -W.a₄ ^ 2 := by
+  rw [WeierstrassCurve.b₈, self.a₁, self.a₂]
+  ring1
+
+theorem c₄ : W.c₄ = -48 * W.a₄ := by
+  rw [WeierstrassCurve.c₄, self.b₂, self.b₄]
+  ring1
+
+theorem c₆ : W.c₆ = -216 * W.b₆ := by
+  rw [WeierstrassCurve.c₆, self.b₂, self.b₄]
+  ring1
+
+theorem Δ : W.Δ = -64 * W.a₄ ^ 3 - 27 * W.b₆ ^ 2 := by
+  rw [WeierstrassCurve.Δ, self.b₂, self.b₄]
+  ring1
+
+variable [CharP R 2]
+
+theorem b₄_of_char_two : W.b₄ = 0 := by
+  rw [self.b₄]
+  linear_combination W.a₄ * CharP.cast_eq_zero R 2
+
+theorem b₈_of_char_two : W.b₈ = W.a₄ ^ 2 := by
+  rw [self.b₈]
+  linear_combination -W.a₄ ^ 2 * CharP.cast_eq_zero R 2
+
+theorem c₄_of_char_two : W.c₄ = 0 := by
+  rw [self.c₄]
+  linear_combination -24 * W.a₄ * CharP.cast_eq_zero R 2
+
+theorem c₆_of_char_two : W.c₆ = 0 := by
+  rw [self.c₆]
+  linear_combination -108 * W.b₆ * CharP.cast_eq_zero R 2
+
+theorem Δ_of_char_two : W.Δ = W.a₃ ^ 4 := by
+  rw [self.Δ, W.b₆_of_char_two]
+  linear_combination (-32 * W.a₄ ^ 3 - 14 * W.a₃ ^ 4) * CharP.cast_eq_zero R 2
+
+end IsCharTwoJZeroNF
+
+namespace IsCharTwoJZeroNF
+
+variable {E} (self : E.IsCharTwoJZeroNF)
+include self
+
+theorem j_of_char_two [CharP F 2] : E.j = 0 := by
+  simp [EllipticCurve.j, self.c₄_of_char_two]
+
+end IsCharTwoJZeroNF
+
+/-! ### Normal forms of characteristic = 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 2, if it is
+$Y^2 + XY = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharTwoJNeZeroNF`) or
+$Y^2 + a_3Y = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsCharTwoJZeroNF`). -/
+def IsCharTwoNF : Prop := W.IsCharTwoJNeZeroNF ∨ W.IsCharTwoJZeroNF
+
+section VariableChange
+
+variable [CharP R 2] [CharP F 2]
+
+/-- For a `WeierstrassCurve` defined over a ring of characteristic = 2,
+there is an explicit change of variables, which changes it to $Y^2 + a_3Y = X^3 + a_4X + a_6$
+(`WeierstrassCurve.IsCharTwoJZeroNF`) if its j = 0. -/
+def toCharTwoJZeroNF : VariableChange R := ⟨1, W.a₂, 0, 0⟩
+
+/-- For a `WeierstrassCurve` defined over a field of characteristic = 2,
+there is an explicit change of variables, which changes it to $Y^2 + XY = X^3 + a_2X^2 + a_6$
+(`WeierstrassCurve.IsCharTwoJNeZeroNF`) if its j ≠ 0. -/
+def toCharTwoJNeZeroNF (W : WeierstrassCurve F) (ha₁ : W.a₁ ≠ 0) : VariableChange F :=
+  ⟨Units.mk0 _ ha₁, W.a₃ / W.a₁, 0, (W.a₁ ^ 2 * W.a₄ + W.a₃ ^ 2) / W.a₁ ^ 3⟩
+
+theorem toCharTwoJZeroNF_spec (ha₁ : W.a₁ = 0) :
+    (W.variableChange W.toCharTwoJZeroNF).IsCharTwoJZeroNF := by
+  constructor
+  · simp [toCharTwoJZeroNF, ha₁]
+  · simp [toCharTwoJZeroNF, show W.a₂ + 3 * W.a₂ = 0 by
+      linear_combination 2 * W.a₂ * CharP.cast_eq_zero R 2]
```

Comment:
```suggestion
  · simp [toCharTwoJZeroNF]
    linear_combination 2 * W.a₂ * CharP.cast_eq_zero R 2
```

Reply:
But this will introduce a non-terminating `simp`, which is discouraged in mathlib if I remembered correctly.

Reply:
Sure, but in many places you have `field_simp` followed by `linear_combination`, and if I recall correctly `field_simp` calls all of `simp` in addition to those specifically tagged with `field_simp`.

Reply:
I think a compromise is write `simp_rw [toCharTwoJEqZeroNF, variableChange_a₂, inv_one, Units.val_one]` instead of `simp [toCharTwoJZeroNF]`.

Reply:
I just found [this post](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60simp.60.20followers) where they discuss what's OK after a `simp`, and I'd argue that `linear_combnation` is basically a `ring`.

Reply:
I also just found [this](https://github.com/leanprover-community/mathlib4/blob/b758def5262ecb106a626ec885176ed1ece2e4ba/scripts/lint-style.py#L154) which are the tactics that are OK after a `simp`, which includes `ring`.

Reply:
> I also just found [this](https://github.com/leanprover-community/mathlib4/blob/b758def5262ecb106a626ec885176ed1ece2e4ba/scripts/lint-style.py#L154) which are the tactics that are OK after a `simp`, which includes `ring`.

Interesting. But that is not including `linear_combination`...

Reply:
Yeah, and I'm advocating for it to be included on Zulip right now :P

Reply:
The consensus so far is that `field_simp` shouldn't be used before `linear_combination` (and might be added to the flexible list soon); instead we should use `field_simp only` which perhaps someone is working on.

Reply:
> The consensus so far is that `field_simp` shouldn't be used before `linear_combination` (and might be added to the flexible list soon); instead we should use `field_simp only` which perhaps someone is working on.

Oh, that's a bad news... I remembered that `field_simp; ...` is also used in e.g. `Weierstrass.lean` and other files. What about `suffices ...; field_simp ...`?

In my opinion the default `field_simp` should only do `ring` + `field_simp` lemmas. But I think non-terminal `ring` is also discouraged, right?

Reply:
Yeah it's bad news but I do agree with it somewhat, and non-terminal `ring` is indeed discouraged. To be honest `Weierstrass.lean`, `Affine.lean`, and `Group.lean` were written a long time ago and I have since learnt how to write better Lean. For now we don't have `field_simp only`, so let's just keep it as you have written.